Kernel/Net/NE2000NetworkAdapter.cpp

```diff
@@ -0,0 +1,487 @@
+static constexpr int NE2K_RAM_SEND_END = 16384 + 6 * 256;
+static constexpr int NE2K_RAM_SEND_SIZE = NE2K_RAM_SEND_END - NE2K_RAM_SEND_BEGIN;
+
+static constexpr int NE2K_RAM_RECV_BEGIN = NE2K_RAM_SEND_END;
+static constexpr int NE2K_RAM_RECV_END = NE2K_RAM_END;
+static constexpr int NE2K_RAM_RECV_SIZE = NE2K_RAM_RECV_END - NE2K_RAM_RECV_BEGIN;
+
+static_assert(NE2K_RAM_BEGIN % NE2K_PAGE_SIZE == 0);
+static_assert(NE2K_RAM_END % NE2K_PAGE_SIZE == 0);
+static_assert(NE2K_RAM_SEND_BEGIN % NE2K_PAGE_SIZE == 0);
+static_assert(NE2K_RAM_SEND_END % NE2K_PAGE_SIZE == 0);
+static_assert(NE2K_RAM_RECV_BEGIN % NE2K_PAGE_SIZE == 0);
+static_assert(NE2K_RAM_RECV_END % NE2K_PAGE_SIZE == 0);
+
+struct [[gnu::packed]] received_packet_header {
+    u8 status;
+    u8 next_packet_page;
+    u16 length;
+};
+
+void NE2000NetworkAdapter::detect()
+{
+    static const auto ne2k_ids = Array<PCI::ID, 11> {
+        PCI::ID { 0x10EC, 0x8029 }, // RealTek RTL-8029(AS)
+
+        // List of clones, taken from Linux's ne2k-pci.c
+        PCI::ID { 0x1050, 0x0940 }, // Winbond 89C940
+        PCI::ID { 0x11f6, 0x1401 }, // Compex RL2000
+        PCI::ID { 0x8e2e, 0x3000 }, // KTI ET32P2
+        PCI::ID { 0x4a14, 0x5000 }, // NetVin NV5000SC
+        PCI::ID { 0x1106, 0x0926 }, // Via 86C926
+        PCI::ID { 0x10bd, 0x0e34 }, // SureCom NE34
+        PCI::ID { 0x1050, 0x5a5a }, // Winbond W89C940F
+        PCI::ID { 0x12c3, 0x0058 }, // Holtek HT80232
+        PCI::ID { 0x12c3, 0x5598 }, // Holtek HT80229
+        PCI::ID { 0x8c4a, 0x1980 }, // Winbond W89C940 (misprogrammed)
+    };
+    PCI::enumerate([&](const PCI::Address& address, PCI::ID id) {
+        if (address.is_null())
+            return;
+        if (!ne2k_ids.span().contains_slow(id))
+            return;
+        u8 irq = PCI::get_interrupt_line(address);
+        [[maybe_unused]] auto& unused = adopt(*new NE2000NetworkAdapter(address, irq)).leak_ref();
+    });
+}
+
+NE2000NetworkAdapter::NE2000NetworkAdapter(PCI::Address address, u8 irq)
+    : PCI::Device(address, irq)
+    , m_io_base(PCI::get_BAR0(pci_address()) & ~3)
+{
+    set_interface_name("ne2k");
+
+    klog() << "NE2000: Found @ " << pci_address();
+
+    m_interrupt_line = PCI::get_interrupt_line(pci_address());
+    klog() << "NE2000: Port base: " << m_io_base;
+    klog() << "NE2000: Interrupt line: " << m_interrupt_line;
+
+    int ram_errors = ram_test();
+    klog() << "NE2000: RAM test " << (ram_errors > 0 ? "OK" : "KO") << ", got " << ram_errors << " byte errors";
+
+    reset();
+    set_mac_address(m_mac_address);
+    klog() << "NE2000: MAC address: " << m_mac_address.to_string().characters();
+    enable_irq();
+}
+
+NE2000NetworkAdapter::~NE2000NetworkAdapter()
+{
+}
+
+void NE2000NetworkAdapter::handle_irq(const RegisterState&)
+{
+    u8 status = in8(REG_RW_INTERRUPTSTATUS);
+#if NE2000_DEBUG
+    klog() << "NE2000NetworkAdapter: Got interrupt, status=0x" << String::format("%02x", status);
+#endif
```

Comment:
Instead of this construct, please do `dbgln<NE2000_DEBUG>(...)`
That way, the code will not bitrot :)